### ISSUE TITLE
private-name mangling and unresolved reference

### ISSUE THREAD

**Issue by rzwq** [astral-sh/ty#3980]
### Summary

I know, its unexpected behaviour (UB in **Python**, Jesus), but for me interesting to know, does it **ty** bug or not?

Code:

```python

_Human__after_all = 42


class Human:
    def play(self) -> str:
        return __after_all


if __name__ == "__main__":
    print(Human().play())
```

<img width="2380" height="368" alt="Image" src="https://github.com/user-attachments/assets/6bfaf99f-49a2-49b7-ac04-c9b1bbfecd10" />

<img width="2350" height="68" alt="Image" src="https://github.com/user-attachments/assets/8cfa1377-f50c-4e08-a55d-97e2db8d7473" />

### Version

0.0.54

**Comment by MichaReiser** [comment]
This is about Python's private-name mangling (https://docs.python.org/3/reference/expressions.html#private-name-mangling). 

And this is tracked in https://github.com/astral-sh/ty/issues/645

**Comment by rzwq** [comment]
I mean, why **ty** mark this as undefined variable, when its incorrect return type?

**Comment by carljm** [comment]
Because ty (like every other Python type checker) does not model Python's private-name-mangling behavior (which is intended to be an internal implementation detail and not relied on, as this code snippet does). So as far as ty (and every other Python type checker) is concerned, the name `__after_all` is not defined anywhere.

If you want to discuss whether ty should prioritize being the first Python type checker to model this runtime implementation detail, #645 is the place for that discussion.
